### PR TITLE
Setup host & version specifc caches

### PR DESCRIPTION
**Pull request by seancpeters**
probably has a lot of old junk due to git craziness.

### CODE REVIEW

File: src/Microsoft.TemplateEngine.Abstractions/ITemplateEngineHost.cs [hunk 1]
```diff
@@ -11,6 +11,10 @@ public interface ITemplateEngineHost
 
         string HostIdentifier { get; }
 
+        string AssemblyName { get; }
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Abstractions/ITemplateEngineHost.cs]: Let's remove this one

File: src/Microsoft.TemplateEngine.Abstractions/ITemplateEngineHost.cs [hunk 2]
```diff
@@ -11,6 +11,10 @@ public interface ITemplateEngineHost
...
+
+        Version AssemblyVersion { get; }
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Abstractions/ITemplateEngineHost.cs]: Let's rename this one to just `Version`

File: src/Microsoft.TemplateEngine.Utils/EngineEnvironmentSettings.cs [hunk 3]
```diff
@@ -13,10 +13,12 @@ namespace Microsoft.TemplateEngine.Utils
     public static class EngineEnvironmentSettings
     {
         private static readonly string DefaultLocale = "en_US";
+        private static readonly string DefaultAssemblyName = "None";
+        private static readonly Version DefaultAssemblyVersion = new Version("0.0.0.1");
 
         static EngineEnvironmentSettings()
         {
-            Host = new DefaultTemplateEngineHost(string.Empty, DefaultLocale);
+            Host = new DefaultTemplateEngineHost(string.Empty, DefaultLocale, DefaultAssemblyName, DefaultAssemblyVersion);
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Utils/EngineEnvironmentSettings.cs]: This needs to be removed I think. The actual host should do a `EngineEnvironmentSettings.Host = (instance that the host provides here)`

File: src/Microsoft.TemplateEngine.Utils/EngineEnvironmentSettings.cs [hunk 4]
```diff
@@ -57,7 +59,7 @@ public string BaseDir
                 {
                     if (_baseDir == null)
                     {
-                        _baseDir = Path.Combine(UserProfileDir, ".netnew");
+                        _baseDir = Path.Combine(UserProfileDir, ".netnew", Host.AssemblyName, Host.AssemblyVersion.ToString());
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Utils/EngineEnvironmentSettings.cs]: This one should get captured in the renaming/changing of the name/version pair in the host settings

File: src/Microsoft.TemplateEngine.Utils/EngineEnvironmentSettings.cs (hunk 3 above)
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Utils/EngineEnvironmentSettings.cs]: Actually, we already do that in Program.cs, but we shouldn't have the default I don't think. This must always come from the host.

Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Utils/EngineEnvironmentSettings.cs]: Agree, this should go away. It was a cheap way to get around the problem of not always having the locale when the host first needs to be invoked. I added an DefaultTemplateEngineHost.UpdateLocale method to do it more properly.

**Pull request merged**